tests/integration/test_tensorflow_v2_2_savedmodel_artifact.py

```diff
@@ -0,0 +1,140 @@
+import time
+import json
+import urllib
+import pytest
+
+import tensorflow as tf
+
+import bentoml
+from tests.bento_service_examples.tensorflow_classifier import TensorflowClassifier
+
+
+test_data = [[1, 2, 3, 4, 5]]
+test_tensor = tf.constant(test_data)
+
+
+class Tensorflow2Model(tf.keras.Model):
+    def __init__(self):
+        super(Tensorflow2Model, self).__init__()
+        # Simple linear layer which sums the inputs
+        self.dense = tf.keras.layers.Dense(
+            units=1,
+            input_shape=(5,),
+            use_bias=False,
+            kernel_initializer=tf.keras.initializers.Ones(),
+        )
+
+    def call(self, inputs):
+        return self.dense(inputs)
+
+
+@pytest.fixture(scope="module")
+def tf2_svc():
+    """Return a TensorFlow2 BentoService."""
+    # When the ExampleBentoService got saved and loaded again in the test, the
+    # two class attribute below got set to the loaded BentoService class.
+    # Resetting it here so it does not effect other tests
+    TensorflowClassifier._bento_service_bundle_path = None
+    TensorflowClassifier._bento_service_bundle_version = None
+
+    svc = TensorflowClassifier()
+    model = Tensorflow2Model()
+    svc.pack('model', model)
+
+    return svc
+
+
+@pytest.fixture(scope="module")
+def tf2_svc_saved_dir(tmp_path_factory, tf2_svc):
+    """Save a TensorFlow2 BentoService and return the saved directory."""
+    # Must be called at least once before saving so that layers are built
+    # See: https://github.com/tensorflow/tensorflow/issues/37439
+    tf2_svc.predict(test_tensor)
+
+    tmpdir = str(tmp_path_factory.mktemp("tf2_svc"))
+    tf2_svc.save_to_dir(tmpdir)
+
+    return tmpdir
+
+
+@pytest.fixture()
+def tf2_svc_loaded(tf2_svc_saved_dir):
+    """Return a TensorFlow2 BentoService that has been saved and loaded."""
+    return bentoml.load(tf2_svc_saved_dir)
+
+
+@pytest.fixture()
+def tf2_image(tf2_svc_saved_dir):
+    import docker
+
+    client = docker.from_env()
+    image = client.images.build(path=tf2_svc_saved_dir, tag="example_service", rm=True)[
+        0
+    ]
+    yield image
+    client.images.remove(image.id)
+
+
+def _wait_until_ready(_host, timeout, check_interval=0.5):
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        try:
+            if (
+                urllib.request.urlopen(f'http://{_host}/healthz', timeout=0.1).status
```

Comment:
move the `time.sleep(check_interval - 0.1)` to a `finally` block instead?

The container could get into a state where the API server is not up but the frontend server(either gunicorn or marshal server) is responding 500 responses. In that case, this test code will keep sending requests without any wait.

Reply:
My apologies! That is not actually my code. `_wait_until_ready` was re-used directly from:

https://github.com/bentoml/BentoML/blob/bf8c26f90d5fa8f575bdf040cd2e9b5006513f38/tests/integration/api_server/conftest.py#L31-L43

Just so I could get up and running quickly and focus on the TF parts of the tests.

Reply:
There might have been a better way for me to re-use those functions. I don't normally copy and paste like that, but I figured I would have time to refactor given the WIP nature of this PR.

Reply:
@joshuacwnewton got you, would be nice to also fix it there as well, was there any issue importing that function here in this test?

Reply:
The only issue was that it felt strange to import a function from another directory's `conftest.py` plugin.

I had been wondering if there was a better overall structure for these integration tests, but then put those thoughts aside for the time being.

Reply:
It is worth mentioning that `tf2_image` and `tf2_host` are also based on `image` and `host` in `tests/integration/api_server/conftest.py`. 

That is why I was thinking of a larger refactor where `image`/`host` could be used for all of the framework integration tests. But again, that is a much larger change, hehe.

Reply:
Got it, that makes a lot of sense. We should refactor how those code are re-used in our integration tests, potentially moving them to a `docker_test_utils` file. If you wanna submit a refactoring PR for that, feel free to do it!

Reply:
I have made some commits that reference this thread. :slightly_smiling_face:

Reply:
@joshuacwnewton sounds good, is there anything else you are planning to add to this PR for TF2 tests? if not, I will go ahead and merge it

Reply:
Nothing new, I think! Just TF1, but that's better off for another PR. I will edit the title/description to remove mentions of TF1 and then feel free to merge. :)